handbrake: 1.1.2 -> 1.2.0

Anton-Latukha opened the pull request:
###### Motivation for this change
Major update, switch from LibAV to FFMpeg & other changes ([release notes](https://github.com/HandBrake/HandBrake/releases/tag/1.2.0)).
Due to need in HandBrake compilation - `nv-codec-headers` bringed to top-level scope, and during that updated 8.1.24.2 -> 8.2.15.6.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Anton-Latukha committed:
handbrake: 1.1.2 -> 1.2.0

Anton-Latukha commented:
PR awaits: https://github.com/HandBrake/HandBrake/issues/1746, https://github.com/HandBrake/HandBrake/issues/1747, https://github.com/HandBrake/HandBrake/issues/1748 to be solved/sorted-out.

Anton-Latukha committed:
handbrake: 1.2.0: switch to FFMpeg, deps sort

Anton-Latukha committed:
handbrake: 1.2.0: add cmake, no cmake conf

Anton-Latukha committed:
handbrake: 1.2.0: updating deps

rm yasm - dropped
rm bzip2, zlib - no needed for build
rm mp4v2, libmkv, mpeg2dec - drop abandonware, HandBrake now uses FFMpeg for that
add nasm, speex, nv-codec-headers - new required deps

Anton-Latukha commented:
Support of LibAV dropped.

Due to HandBrake having a lot of out-of-tree patches: [folder](https://github.com/HandBrake/HandBrake/tree/master/contrib).

Many of them seem like a good quality improvements. Some of them go upstream, some not.
It is not possible to track all their patches, and if we do - it would be essentially their bundle in the closure.

Some sort of official internal build instructions are:
1. https://github.com/HandBrake/HandBrake/blob/master/pkg/linux/flatpak/fr.handbrake.ghb.json
2. https://github.com/flathub/fr.handbrake.ghb/blob/master/fr.handbrake.ghb.json

External build extructions: https://handbrake.fr/docs/en/1.2.0/developer/build-linux.html

Anton-Latukha commented:
I would be happy to go through all major and minor notes about the code of the package.
I want to learn more.

Anton-Latukha commented:
One of the considerations may be:
Instead of FFMpeg_4: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/ffmpeg/generic.nix. Since FFMpeg in Nix is:
https://github.com/NixOS/nixpkgs/blob/830599debecac99f0215029c922879af84be386e/pkgs/development/libraries/ffmpeg/generic.nix#L23

To use FFMpeg-full: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/ffmpeg-full/default.nix

FFMpeg-full has more codecs/containers for input, which seems sane for the rich video transcoder.

Anton-Latukha commented:
Diff of FFMpeg-{full,4} closures is 100mb.

Anton-Latukha committed:
handbrake: 1.2.0: ffmpeg_4 -> ffmpeg-full

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review-december/1711/7

FRidh commented:
@GrahamcOfBorg eval

peterhoeg commented on pkgs/applications/video/handbrake/default.nix:
```diff
@@ -1,72 +1,70 @@
 
+let
+
+  nv-codec-headers = callPackage ../../../development/libraries/ffmpeg-full/nv-codec-headers.nix { };
```
Why are you using callPackage here when you can simply pass in ```nv-codec-headers``` to the main function?

Anton-Latukha replied:
Because there is no such package?
It is declared in `pkgs/development/libraries/ffmpeg-full/nv-codec-headers.nix`
Not declared at top level, that is why I think there is not attribute.

Since it used internally only in FFMpeg-full and required for building HandBrake:
```
../libhb/nvenc_common.c:13:10: fatal error: ffnvcodec/nvEncodeAPI.h: No such file or directory
 #include <ffnvcodec/nvEncodeAPI.h>
          ^~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [../libhb/module.rules:12: libhb/nvenc_common.o] Error 1
```

That is why I used `nv-codec-headers` the way FFMpeg-full solely uses it, by directly calling a package.

Anton-Latukha replied:
Should I expose `nv-codec-headers` as a package?

Anton-Latukha committed:
nv-codec-headers: 8.1.24.2 -> 8.2.15.6

Anton-Latukha committed:
nv-codec-headers: bring package to top-level

Anton-Latukha commented:
@MP2E
Updated here `handbrake` to `1.2.0` requires `nv-codec-headers` for compilation.
I moved the `nv-codec-headers` to the global NixPkgs file structure and top-level package scope.
The package was used internally only for `ffmpeg-full`.
I also updated `nv-codec-headers` `8.1.24.2` -> `8.2.15.6` during the process.

Anton-Latukha commented:
Everything is done.
It is pretty straight-forward PR.
Please, merge it.
